update-annotations-table.R

```diff
@@ -0,0 +1,101 @@
+################################################
+####  Update the Synapse Annotations table  ####
+################################################
+
+library("jsonlite")
+library("tidyverse")
+library("reticulate")
+library("glue")
+library("httr")
+library("here")
+
+"%||%" <- function(a, b) {
+  if (!is.null(a)) a else b
+}
+
+#' Extract module or annotation key name from $id field
+#'
+#' @param id string containing the schema `$id`
+#' @param info one of "module" or "key"
+#' @return The module or key name
+get_info <- function(id, info = c("module", "key")) {
+  info <- match.arg(info)
+  pattern <- "^[^-]+-([[:alnum:]]+)\\.([[:alnum:]]+)-[0-9\\.]+"
+  switch(
+    info,
+    module = gsub(pattern, "\\1", id),
+    key = gsub(pattern, "\\2", id)
+  )
+}
+
+#' Create table row(s) for a term
+#'
+#' @param file Path to the JSON Schema file containing a term
+#' @return
+create_rows <- function(file) {
+  dat <- fromJSON(file)
+  ## Information about the key
+  return_df <- tibble::tibble(
+    key = get_info(dat$`$id`, "key"),
+    description = dat$description %||% NA_character_,
+    columnType = dat$type %||% "string",
+    module = get_info(dat$`$id`, "module"),
+    maximumSize = NA_real_
+  )
+  ## If term has enumerated values in anyOf, fromJSON will create a handy data
+  ## frame of those. We use that in combination with return_df to create a row
+  ## for each value.
+  if ("anyOf" %in% names(dat)) {
+    enum_df <- dat[["anyOf"]] %>%
+      rename(
+        value = const,
+        valueDescription = description
+      ) %>%
+      mutate(value = as.character(value))
+    return_df <- full_join(return_df, enum_df, by = character())
+  } else {
+    return_df <- return_df %>%
+      add_column(value = NA, valueDescription = NA, source = NA)
+  }
+  ## Select columns in order that matches the table schema
+  select(
+    return_df,
+    key,
+    description,
+    columnType,
+    maximumSize,
+    value,
+    valueDescription,
+    source,
+    module
+  )
+}
+
+## Log in to synapse
+synapse <- import("synapseclient")
+syn <- synapse$Synapse()
+syn$login()
+
+## JSON files
+files <- list.files(
+  here("terms"),
+  pattern = "\\.json$",
+  full.names = TRUE,
+  recursive = TRUE
+)
+
+dat <- map_dfr(files, create_rows)
```

Comment:
Thanks for catching this. I must have created this script from a slightly out of date version of master, before that antibodyOrder PR was merged. Perhaps an argument for the "require up-to-date branches before merging" setting we were talking about yesterday.

Reply:
In the case of scripts, I could definitely see using the 'require up-to-date branches before merging' setting. I'm still not completely sure it's needed for the majority of what will be added (keys/values), though. Regardless, I'm still on board with adding it back in, if it might help keep things on track.